src/jaxsim/exceptions.py

```diff
@@ -0,0 +1,55 @@
+import jax
+
+
+def raise_if(
+    condition: bool | jax.Array, exception: type, msg: str, *args, **kwargs
+) -> None:
+    """
+    Raise a host-side exception if a condition is met. Useful in jit-compiled functions.
+
+    Args:
+        condition:
+            The boolean condition of the evaluated expression that triggers
+            the exception during runtime.
+        exception: The type of exception to raise.
+        msg:
+            The message to display when the exception is raised. It can be a fmt string,
+            and users can pass additional arguments to format the string.
+    """
+
+    # Check early that the fmt string is well-formed.
+    _ = msg.format(*args, **kwargs)
+
+    def _raise_exception(condition: bool, *args, **kwargs) -> None:
+        if condition:
+            raise exception(msg.format(*args, **kwargs))
+
+    def _callback(args, kwargs) -> None:
+        jax.debug.callback(_raise_exception, condition, *args, **kwargs)
+
+    # Since running a callable on the host is expensive, we prevent its execution
+    # if the condition is False with a low-level conditional expression.
+    def _run_callback_only_if_condition_is_true(*args, **kwargs) -> None:
+        return jax.lax.cond(
+            condition,
+            _callback,
+            lambda args, kwargs: None,
+            args,
+            kwargs,
+        )
+
+    return _run_callback_only_if_condition_is_true(*args, **kwargs)
```

Comment:
We can avoid calling `format` when the condition is false. Therefore avoiding to pass `args` and `kwargs` to the other functions
```suggestion
    try:
        exc_msg = msg.format(*args, **kwargs)
    except Exception as e:
        raise ValueError("Error in formatting exception message: " + str(e))

    def _raise_exception() -> None:
        raise exception(exc_msg)

    def _callback() -> None:
        jax.debug.callback(_raise_exception)

    def _run_callback_only_if_condition_is_true() -> None:
        return jax.lax.cond(
            condition, lambda: _callback(), lambda: None,
        )

    # Return the callback function to be used within a JIT-compiled context.
    return _run_callback_only_if_condition_is_true()
```

Reply:
This won't work. If you don't pass properly `args` and `kwargs` as it is done in this PR, the returned function is closed over the _traced_ `args` and `kwargs`. You won't get the values produced during runtime within a jit-compiled function.

Reply:
Regardless, I introduces as suggested a try clause that provides useful information for debugging purpose (even if it may fail only on JAX traces) like the following:

```
ValueError: Error in formatting exception message with args=() and kwargs={'numm': Traced<ShapedArray(int64[], weak_type=True)>with<DynamicJaxprTrace(level=1/0)>}
```

Reply:
When I tried that it worked with the test you provided. In which case do you think it may fail?

Reply:
I don't understand why I get access to the runtime value of data instead:

<details><summary>Details</summary>
<p>

```python
Compiling...
2024-06-18 14:39:29.711244: W external/xla/xla/service/gpu/runtime/support.cc:58] Intercepted XLA runtime error:
INTERNAL: CpuCallback error: ValueError: This is a test exception for 42 data

At:
  /home/flferretti/jaxsim/src/jaxsim/exceptions.py(31): _raise_exception
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(246): _flat_callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(89): debug_callback_impl
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(153): _callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/interpreters/mlir.py(2447): _wrapped_callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py(1201): __call__
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/profiler.py(336): wrapper
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1348): _pjit_call_impl_python
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1392): call_impl_cache_miss
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1415): _pjit_call_impl
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(913): process_primitive
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(423): bind_with_trace
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(2727): bind
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(143): _python_pjit_helper
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(248): cache_miss
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/traceback_util.py(179): reraise_with_filtered_traceback
  <ipython-input-1-919b7cc979d7>(23): <module>
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3577): run_code
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3517): run_ast_nodes
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3334): run_cell_async
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/async_helpers.py(129): _pseudo_sync_runner
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3130): _run_cell
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3075): run_cell
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/interactiveshell.py(904): interact
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/interactiveshell.py(911): mainloop
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/ipapp.py(317): start
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/traitlets/config/application.py(1075): launch_instance
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/__init__.py(130): start_ipython
  /home/flferretti/miniforge3/envs/jaxsim/bin/ipython(10): <module>

2024-06-18 14:39:29.711308: E external/xla/xla/pjrt/pjrt_stream_executor_client.cc:2732] Execution of replica 0 failed: INTERNAL: Failed to execute XLA Runtime executable: run time error: custom call 'xla.gpu.custom_call' failed: CpuCallback error: ValueError: This is a test exception for 42 data

At:
  /home/flferretti/jaxsim/src/jaxsim/exceptions.py(31): _raise_exception
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(246): _flat_callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(89): debug_callback_impl
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/debugging.py(153): _callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/interpreters/mlir.py(2447): _wrapped_callback
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py(1201): __call__
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/profiler.py(336): wrapper
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1348): _pjit_call_impl_python
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1392): call_impl_cache_miss
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(1415): _pjit_call_impl
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(913): process_primitive
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(423): bind_with_trace
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/core.py(2727): bind
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(143): _python_pjit_helper
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/pjit.py(248): cache_miss
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/jax/_src/traceback_util.py(179): reraise_with_filtered_traceback
  <ipython-input-1-919b7cc979d7>(23): <module>
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3577): run_code
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3517): run_ast_nodes
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3334): run_cell_async
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/async_helpers.py(129): _pseudo_sync_runner
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3130): _run_cell
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/core/interactiveshell.py(3075): run_cell
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/interactiveshell.py(904): interact
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/interactiveshell.py(911): mainloop
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/terminal/ipapp.py(317): start
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/traitlets/config/application.py(1075): launch_instance
  /home/flferretti/miniforge3/envs/jaxsim/lib/python3.11/site-packages/IPython/__init__.py(130): start_ipython
  /home/flferretti/miniforge3/envs/jaxsim/bin/ipython(10): <module>
; current tracing scope: custom-call.58; current profiling annotation: XlaModule:#prefix=jit(jit_compiled_function)/jit(main),hlo_module=jit_jit_compiled_function,program_id=30#.
```

</p>
</details>

Reply:
Me neither. Can you try running the snippet above after checking out https://github.com/ami-iit/jaxsim/tree/test_exception_with_fil? I still get the traced value there.

Reply:
Yep, my bad I was using the wrong commit

Reply:
No problem, glad that things work in the same way on both our setups :)